### PR TITLE
[JDBC-V2] Fix Parser Issues

### PR DESCRIPTION
**Pull request by chernser**
## Summary

This PR introduces option to select old JavaCC parser that is proven to work and has more relaxed rules.  Parameter parsing code is ported from older version as most stable logic. And this PR will make old parser as choice by default so it will work almost like JDBC V1. 

And: 
- Fixes CTE issue 
- Fixes issue with using keyword as alias
- Fixes problem when query falsely detected as one without result set 
- Fixes issue with logging 
- Fixes using TIMESTAMP in queries

Closes https://github.com/ClickHouse/clickhouse-java/issues/2574
Closes https://github.com/ClickHouse/clickhouse-java/issues/2568
Closes https://github.com/ClickHouse/clickhouse-java/issues/2537
Closes https://github.com/ClickHouse/clickhouse-java/issues/2595
Closes https://github.com/ClickHouse/clickhouse-java/issues/2617

Closes https://github.com/ClickHouse/clickhouse-java/issues/2569
Closes https://github.com/ClickHouse/clickhouse-java/issues/2570
Closes https://github.com/ClickHouse/clickhouse-java/issues/2571
Closes https://github.com/ClickHouse/clickhouse-java/issues/2572
Closes https://github.com/ClickHouse/clickhouse-java/issues/2573
Closes https://github.com/ClickHouse/clickhouse-java/issues/2609

## Checklist
Delete items not relevant to your PR:
- [x] Closes #<issue ref>
- [x] Unit and integration tests covering the common scenarios were added
- [ ] A human-readable description of the changes was provided to include in CHANGELOG
- [ ] For significant changes, documentation in https://github.com/ClickHouse/clickhouse-docs was updated with further explanations or tutorials

### CODE REVIEW

File: jdbc-v2/src/test/java/com/clickhouse/jdbc/internal/BaseSqlParserFacadeTest.java [hunk 1]
```diff
@@ -233,22 +233,42 @@ public static Object[][] testCreateStmtDP() {
...
+                {"WITH toDateTime('2025-08-20 12:34:56') AS target_time SELECT * FROM table", 0},
+                {"WITH toDate('2025-08-20') as DATE_END, events AS ( SELECT 1 ) SELECT * FROM events", 0},
+                {"WITH toDate(?) as DATE_END, events AS ( SELECT 1 ) SELECT * FROM events", 1},
+                {"WITH ? as a, ? as b, body as ( select ? ) select a, b, body.* from body", 3},
+                {"WITH 'a_value' as a, 'b_value' as b, body as ( select 'html' ) select a, b, body.* from body", 0},
+                {"WITH 'a_value' as a, 'b_value' as b, body as ( with 'data' as d select d, 'html' ) select a, b, body.* from body", 0},
+                {"with date select date, 1 from (select now() date)", 0 },
+                {COMPLEX_CTE, 4},
+                {"WITH 'date' as const1, 'time' as const2, Tmp1 as (SELECT 1), Tmp2 as (SELECT * FROM Tmp1) SELECT * FROM Tmp2 ", 0},
+                {"WITH query1 AS ( WITH 'a' as date1 SELECT * FROM tracking.event WHERE project='a' AND time>=starting_time AND time<ending_time GROUP BY date, user_id ) SELECT * FROM query1", 0},
```
Comment (AVMusorin) [inline jdbc-v2/src/test/java/com/clickhouse/jdbc/internal/BaseSqlParserFacadeTest.java:227]: Could you please add following tests, they will cover position swapping for CSE and CTE that I described in https://github.com/ClickHouse/clickhouse-java/pull/2611:

- `{"with a as (select ?), (select 2) as b select b, *, c from a;", 1 }`
- `{ "with a as (select ?), (select 2) as b, c as (select ?) select *, b from a, c;", 2 }`
- `{ "with (select 2) as b, a as (select ?), (select 3) as c select *, b, c from a;", 1 },`

Comment (chernser) [reply jdbc-v2/src/test/java/com/clickhouse/jdbc/internal/BaseSqlParserFacadeTest.java:227]: Thank you for reminder! 
Yes, will add it.

File: jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/ClickHouseParser.g4 [hunk 2]
```diff
@@ -66,9 +63,9 @@ alterStmt
     ;
 
 alterTableClause
-    : ADD COLUMN (IF NOT EXISTS)? tableColumnDfnt (AFTER nestedIdentifier)?         # AlterTableClauseAddColumn
-    | ADD INDEX (IF NOT EXISTS)? tableIndexDfnt (AFTER nestedIdentifier)?           # AlterTableClauseAddIndex
-    | ADD PROJECTION (IF NOT EXISTS)? tableProjectionDfnt (AFTER nestedIdentifier)? # AlterTableClauseAddProjection
+    : ADD COLUMN (IF NOT EXISTS)? tableColumnDfnt (AFTER (nestedIdentifier | FIRST))?         # AlterTableClauseAddColumn
```
Comment (EastLord) [inline jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/ClickHouseParser.g4]: Hello
The current rule:
... (AFTER (nestedIdentifier | FIRST))?

This seems to incorrectly allow the syntax AFTER FIRST.

I believe the correct grammar should be:
... ((AFTER nestedIdentifier) | FIRST)?

This change correctly treats FIRST and AFTER nestedIdentifier as two separate, mutually exclusive options, which aligns with the official ClickHouse SQL syntax.

Thanks!

https://clickhouse.com/docs/sql-reference/statements/alter/column

Comment (chernser) [reply jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/ClickHouseParser.g4]: Thank you, @EastLord! 

Your are absolutely correct!

File: jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/antlr4/ClickHouseParser.g4 [hunk 3]
```diff
@@ -432,14 +469,35 @@ withClause
...
     ;
 
+// CTE statement
+cteClause
+    : WITH cteUnboundCol? (COMMA cteUnboundCol)* COMMA? namedQuery? (COMMA namedQuery)*
+    ;
+
+
+namedQuery
+    : identifier (columnAliases)? AS  LPAREN? ( selectStmt | selectStmtWithParens | selectUnionStmt) RPAREN?
```
[the remark this answers is not in the document]
Comment (chernser) [reply jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/antlr4/ClickHouseParser.g4:492]: May be. But it works :-)

File: jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/ClickHouseParser.g4 [hunk 4]
```diff
@@ -66,9 +63,9 @@ alterStmt
     ;
 
 alterTableClause
-    : ADD COLUMN (IF NOT EXISTS)? tableColumnDfnt (AFTER nestedIdentifier)?         # AlterTableClauseAddColumn
-    | ADD INDEX (IF NOT EXISTS)? tableIndexDfnt (AFTER nestedIdentifier)?           # AlterTableClauseAddIndex
-    | ADD PROJECTION (IF NOT EXISTS)? tableProjectionDfnt (AFTER nestedIdentifier)? # AlterTableClauseAddProjection
+    : ADD COLUMN (IF NOT EXISTS)? tableColumnDfnt ((AFTER nestedIdentifier) | FIRST)?         # AlterTableClauseAddColumn
+    | ADD INDEX (IF NOT EXISTS)? tableIndexDfnt ((AFTER nestedIdentifier) | FIRST)?           # AlterTableClauseAddIndex
+    | ADD PROJECTION (IF NOT EXISTS)? tableProjectionDfnt (AFTER (nestedIdentifier | FIRST))? # AlterTableClauseAddProjection
```
Comment (EastLord) [inline jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/ClickHouseParser.g4]: ((AFTER nestedIdentifier) | FIRST)?

File: jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/antlr4/ClickHouseParser.g4 [hunk 5]
```diff
@@ -830,24 +925,90 @@ systemPrivilege
...
+    | SYSTEM RELOAD FUNCTIONS clusterClause?
+    | SYSTEM RELOAD FUNCTION clusterClause? identifier
+    | SYSTEM RELOAD ASYNCHRONOUS METRICS clusterClause?
+    | SYSTEM DROP DNS CACHE
+    | SYSTEM DROP MARK CACHE
+    | SYSTEM DROP REPLICA literal (FROM SHARD literal)? (FROM (TABLE tableIdentifier) | (FROM DATABASE identifier) | (ZKPATH literal))?
+    | SYSTEM DROP UNCOMPRESSED CACHE
+    | SYSTEM DROP COMPILED EXPRESSION CACHE
+    | SYSTEM DROP QUERY CONDITION CACHE
+    | SYSTEM DROP QUERY CACHE (TAG literal)?
```
Comment (EastLord) [inline jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/antlr4/ClickHouseParser.g4:989]: QUERY_SQL or QUERY   ?

File: jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/ClickHouseParser.g4
Comment (EastLord) [inline jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/ClickHouseParser.g4:881]: Hello,
In ClickHouseLexer.g4, two distinct tokens are defined:

QUERY        : '?';
QUERY_SQL    : Q U E R Y; 

However, in ClickHouseParser.g4, the rule for SYSTEM DROP ... CACHE uses the QUERY token:

SYSTEM DROP QUERY CACHE

The parser rule should be updated to use the QUERY_SQL token, which correctly represents the literal keyword.

The line should be changed from:
SYSTEM DROP QUERY CACHE

To:
SYSTEM DROP QUERY_SQL CACHE

This will ensure the grammar correctly parses the valid ClickHouse SQL syntax.

Comment (chernser) [reply jdbc-v2/src/main/antlr4/com/clickhouse/jdbc/internal/parser/ClickHouseParser.g4:881]: Yes, you are right. 
Thank you!

Seems I need to rename query to `JDBC_PARAM`

**Review by EastLord — APPROVED** [review]

**Comment by EastLord** [comment]
Hello.
While reviewing the parser, I've noticed that it is quite comprehensive. However, like any complex grammar, there may still be some edge cases or less common syntax variations that are not yet fully covered.

To help identify and address any remaining gaps, I would like to propose leveraging the extensive suite of test queries directly from the official ClickHouse repository. These queries are used to test the native ClickHouse parser and represent a massive collection of valid (and invalid) SQL statements.

The official test suite can be found here: https://github.com/ClickHouse/ClickHouse/tree/master/tests/queries

**Comment by chernser** [comment]
Good day, @EastLord ! 

Many thanks to your help and suggestions! 

yes, we have already run tests from this suite.  However it requires porting tests to Java, there are no JDBC parameter tests   what adds another dimension for tests. 
Currently we discussing porting old parser to unblock migration to V2 for our users. 
All work would be done in this PR.

File: jdbc-v2/src/main/java/com/clickhouse/jdbc/DriverProperties.java [hunk 6]
```diff
@@ -57,6 +57,17 @@ public enum DriverProperties {
...
+    /**
+     * Configures what SQL parser will be used. Choices:
+     * <ul>
+     *     <li>ANTLR4 - parser extracts required information but PreparedStatement parameters parsed separately.</li>
+     *     <li>ANTLR4_PARAMS_PARSER - parser extracts required information AND parameter positions.</li>
+     *     <li>JAVACC - parser extracts required information but PreparedStatement parameters parsed separately.</li>
+     * </ul>
+     */
+//    SQL_PARSER("jdbc_sql_parser", "JAVACC", List.of("ANTLR4", "ANTLR4_PARAMS_PARSER", "JAVACC")),
+    SQL_PARSER("jdbc_sql_parser", "ANTLR4", List.of("ANTLR4", "ANTLR4_PARAMS_PARSER", "JAVACC")),
```
Comment (mzitnik) [inline jdbc-v2/src/main/java/com/clickhouse/jdbc/DriverProperties.java]: Setting to ANTLR4 means that we are using ANTLR4 as the default one?

**Review by mzitnik — COMMENTED** [review]
There are 2 failing tests 
- com.clickhouse.client.ParameterizedQueryTest.testParamsWithInstant
- PreparedStatementTest.testMetabaseBug01

File: jdbc-v2/src/main/java/com/clickhouse/jdbc/DriverProperties.java (hunk 6 above)
Comment (chernser) [reply jdbc-v2/src/main/java/com/clickhouse/jdbc/DriverProperties.java]: Thank you for catching it! 
Yes, we need to use JAVACC by default.

**Comment by chernser** [comment]
@mzitnik 
I will address these issues in another PR. 
They are failing for some reason
